design/one-pager-generic-usage-type.md

```diff
@@ -0,0 +1,397 @@
+- Preventing deletion of `Usage` before the using resource is deleted.
+- Adding or removing admission webhook rules to intercept the `DELETE` requests
+  for the used resource(s).
+
+**Resolving selectors:**
+
+The API will support defining selectors for both the using and used resources.
+Initially, we will only support resolving references with the following rules:
+
+- Resolution will be made once when the `Usage` is created.
+- If multiple resources match the selector, a random one will be
+  selected.
+
+The [existing policies] for resolving Managed Resource references does not
+necessarily apply to the `Usage`. For example, supporting an `optional`
+reference while defining a usage relationship does not make sense. Also, the
+[reasoning behind] supporting different policies for MRs does not apply to
+the `Usage`. However, depending on the feedback, we may consider supporting
+resolve policies to configure whether the resolution should be made once or
+continuously in a future iteration.
+
+**Adding owner reference:**
+
+We want the `Usage` to be deleted when the using resource is deleted. This will
+be achieved by adding an owner reference from the using resource to the `Usage`.
+
+**Preventing early deletion of the `Usage`:**
+
+Typically, we expect a `Usage` resource to be defined as part of the same
+composition as the using and used resources. This means the `Usage` will get
+the delete request when the composite resource is deleted. In this case, the
+controller will prevent the deletion of the `Usage` until the using resource is
+deleted with the help of a finalizer.
+
+**Defining admission webhook rules:**
+
+The admission webhook will intercept the `DELETE` requests for the used
+resource(s). To achieve this, the controller will add [webhook rules] for the
+used resource(s) when a `Usage` is created. It is also the controller's job to
+remove these rules when the `Usage` is deleted, and no other `Usage` resources
+reference the same used resource(s).
+
+#### Admission Webhook
+
+The admission webhook will be responsible for intercepting the `DELETE` requests
+for the used resource(s) and rejecting them with a ["409 Conflict" error code].
+
+The webhook will be implemented as a Kubernetes validating admission webhook
+leveraging the [existing webhook machinery] in Crossplane and will be running
+as part of the Crossplane core.
+
+We will use the [Field Indexers] to efficiently query the used resource(s) when
+a `DELETE` request is received.
+
+Implementation of the core logic for the webhook handler:
+
+```go
+func (h *handler) validateNoUsages(ctx context.Context, u *unstructured.Unstructured) admission.Response {
+	usageList := &v1alpha1.UsageList{}
+	if err := h.reader.List(ctx, usageList, client.MatchingFields{inUseIndexKey: getIndexValueForObject(u)}); err != nil {
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
+	if len(usageList.Items) > 0 {
```

Comment:
Not sure it's fair to nitpick code in a design, but this looks like a good place to [return early](https://github.com/crossplane/crossplane/tree/master/contributing#return-early). 😉

Reply:
I believe we are already returning early; the return statement for the error case looks bigger than the success below.
Or, am I missing sth?

Reply:
You have a point. I was thinking we could reduce nesting by doing:

```go
if len(usageList.Items) == 0 {
	return admission.Allowed("")
}

return admission.Response{}  // etc..
```

We can close this thread though and worry about my code nitpicks at implementation time. 😉